dolphin/api/v1/alert.py

```diff
@@ -0,0 +1,153 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from oslo_log import log
+from webob import exc
+
+from dolphin import db, cryptor
+from dolphin import exception
+from dolphin import utils
+from dolphin.api import validation
+from dolphin.api.common import wsgi
+from dolphin.api.schemas import alert as schema_alert
+from dolphin.api.views import alert as alert_view
+from dolphin.drivers import api as driver_api
+from dolphin.i18n import _
+
+LOG = log.getLogger(__name__)
+
+
+SNMPv3_keys = ('username', 'password', 'auth_enable', 'auth_protocol', 'encryption_enable',
+               'encryption_protocol', 'encryption_password', 'engine_id')
+
+
+class AlertController(wsgi.Controller):
+    def __init__(self):
+        super().__init__()
+        self.driver_api = driver_api.API()
+
+    @wsgi.response(200)
+    @validation.schema(schema_alert.put)
+    def put(self, req, id, body):
+        """Create a new alert source or update an exist one."""
+        ctx = req.environ['dolphin.context']
+        alert_source = body
+
+        try:
+            alert_source["storage_id"] = id
+            db.storage_get(ctx, id)
+            alert_source = self._input_check(alert_source)
+
+            if self._exist(ctx, id):
+                alert_source = db.alert_source_update(ctx, id, alert_source)
+            else:
+                alert_source = db.alert_source_create(ctx, alert_source)
+        except exception.StorageNotFound as e:
+            msg = (_("Alert source cannot be created or updated for a non-exist storage %s.") % id)
+            raise exc.HTTPBadRequest(explanation=msg)
+        except exception.InvalidInput as e:
+            msg = _('Failed to put alert_source: {0}'.format(e))
+            raise exc.HTTPBadRequest(explanation=msg)
+
+        return alert_view.build_alert_source(self._decrypt_password(alert_source))
+
+    @wsgi.response(200)
+    def show(self, req, id):
+        ctx = req.environ['dolphin.context']
+        try:
+            alert_source = db.alert_source_get(ctx, id)
+        except exception.AlertSourceNotFound as e:
+            raise exc.HTTPNotFound(explanation=e.msg)
+        except Exception as e:
+            LOG.error(e)
+            raise exc.HTTPBadRequest(explanation=e.msg)
+
+        return alert_view.build_alert_source(self._decrypt_password(alert_source))
+
+    @wsgi.response(200)
+    def delete(self, req, id):
+        ctx = req.environ['dolphin.context']
+        try:
+            db.alert_source_delete(ctx, id)
+        except exception.AlertSourceNotFound as e:
+            raise exc.HTTPNotFound(explanation=e.msg)
+
+    def _input_check(self, alert_source):
+        version = alert_source.get('version')
+
+        if version.lower() == 'snmpv3':
+            user_name = alert_source.get('username', None)
+            auth_enabled_str = alert_source.get('auth_enabled', None)
+            engine_id = alert_source.get('engine_id', None)
```

Comment:
If user selects v3 and gives just username and enginid, will it be success. Is it better to consider default auth_enabled_str as false

Reply:
In that case, it will failed and the error message will tell the user auth_enabled is required. Cauz IMO, SNMPv3 is mainly for security, if auth is not enabled, then that will not be safe, so we'd better let the user to explicitly enable or disable authentication.